torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.h

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+#include <torch/csrc/jit/codegen/cuda/python_frontend/fusion_owner.h>
+
+using NvfDataType = torch::jit::fuser::cuda::DataType;
+using NvfTensorView = torch::jit::fuser::cuda::TensorView;
+using NvfVal = torch::jit::fuser::cuda::Val;
+
+namespace nvfuser {
+
+struct RecordFunctor;
+
+enum class StateType {
+  Tensor,
+  Scalar,
+  None,
+};
+
+struct State {
+  State(StateType _stype, size_t _index) : stype(_stype), index(_index) {}
+
+  StateType stype;
+  size_t index;
+};
+
+struct Tensor : State {
+  Tensor(size_t _index) : State(StateType::Tensor, _index) {}
+};
+
+struct Scalar : State {
+  Scalar(size_t _index) : State(StateType::Scalar, _index) {}
+};
+
+// Manually applying the fusion guard via a context manager
+class FusionDefinition {
```

Comment:
Would also be helpful to have some comment here describing how `FusionDefinition` will be used.

Reply:
Is it a `FusionCache` that keeps multiple fusion graphs?

Reply:
Currently, there is limited caching.  The `FusionOwner` is holding an instance of the `FusionExecutorCache` that can keep multiple kernels depending on the inputs for a given Fusion IR object.  A true graph cache is yet to be added once the Recording system is added.  The current implementation always creates a new Fusion IR Fusion object.  The caching is the next step that will be added in a future PR.